Enhancement/cleanup sequences

junniest opened the pull request.

junniest committed:
Add Index and Index mut impls to Sequences

Added Index and Index mut implementations to the Sequences struct to allow
direct iteration over elements and replace obsolete `record` and `record_mut`
methods.

junniest committed:
Remove redundant record access

junniest committed:
Change map visibility in PairwiseAlignment

Changed map_x and pam_y visibility to pub(crate) in PairwiseAlignment, added accessors

junniest committed:
Add len() method to PairwiseAlignment

Added len() method to PairwiseAlignment, now using alignment.len() directly instead of
map_x.len() in tests

junniest committed:
Fix tests missing &

junniest committed:
Add docstrings to Sequences struct

junniest committed:
Implement proper iteration over `Sequences`

junniest committed:
Fix PartielEq for Sequences to remove cloning

junniest committed:
Fix ExactSizeIterator lifetime for clippy

Fixed ExactSizeIterator lifetime for clippy which wanted the lifetime to be elided

MattesMrzik commented on phylo/src/alignment/mod.rs:
```diff
@@ -30,15 +30,28 @@ pub type SeqMaps = HashMap<NodeIdx, Mapping>;
 
 impl PairwiseAlignment {
     pub fn new(map_x: Mapping, map_y: Mapping) -> PairwiseAlignment {
         debug_assert!((map_x.len() == map_y.len()) || map_y.is_empty());
```
Why do we allow the `map_y` to be empty?
Perhaps do this check always and not only in debug, or alternatively trim the longer seq to the len of the shorter.

junniest replied:
Good catch. I actually noticed this but then thought it was for an implementation detail that used to exist but not anymore.

If I remember correctly I used to use the `PairwiseAlignment` struct to also store leaf maps, which would only have one map filled out, the second would be empty. At the moment, however, the leaf maps are stored just as `Mapping`s, so this check is just plain wrong. Removed it, it only checks that the maps match in length now.

To your second comment:
The check is debug because I tend to avoid assertions in production code for performance reasons, which is not the case here. This one doesn't seem to have much of a cost attached and it's part of the public API, so it should actually be a regular assert, not a debug. Plus in Rust it's acceptable for `new` to panic, so if someone tries to create an invalid `PairwiseAlignment` a panic is in order.

But it also should never be the case that something needs to be trimmed, the two maps must match in length. Trimming something will mean that we are discarding information and that could hide some nasty bugs. So panic it is!

MattesMrzik commented on phylo/src/alignment/sequences.rs:
````diff
@@ -60,34 +140,87 @@ impl Sequences {
+    ///     println!("{}", record.id());
+    /// }
+    /// ```
+    pub fn iter(&self) -> Iter<'_> {
````
Would it also be possible to return `slice::Iter<'a, Record>` here? Then the code for `Iter`, ... would also not be needed.

junniest replied:
Good catch! I ended up implementing the iterators using CoPilot and so paid for it :D I didn't know how to implement iterators properly, but the fluff with `Iter` is clearly unnecessary, thank you.

junniest committed:
Fix typo in doctring

junniest committed:
Remove incorrect check for `map_y.is_empty()`

Removed incorrect check for `map_y.is_empty()` in `PairwiseAlignment::new()`.
This was a rudiment of a previous implementation detail where PairwiseAlignment
was used to store leaf maps with an empty `map_y`, which is no longer the case.

junniest committed:
Change `debug!` to `assert!` for map lengths

Changed `debug_assert!` to regular `assert_eq!` for comparing map lengths when creating
a new `PairwiseAlignment` and added an error message. The two maps must be of equal length upon
creation of the struct and this is not a recoverable error, thus production code should panic.

junniest committed:
Add debug assert for map lengths in `len()`

Added `debug_assert_eq!` for map lengths in `PairwiseAlignment::len()`.
Debug because it is impossible to create an alignment with unequal map lenghts
with the public API.

junniest committed:
Simplify iterator impl over `Sequences`

Simplified the iterator implementation over `Sequences`, removed unnecessary
additional struct.

junniest committed:
Add more tests for `Sequences::eq`

junniest committed:
Remove "test" from test names

junniest committed:
Add tests for record access and mut access

junniest committed:
Fix iterator bug and added test

junniest committed:
Make `Sequences::iter()` private

Made `Sequences::iter()` private and replaced its usage it tests

junniest committed:
Remove doctring from private `Sequences::iter()`

junniest committed:
Expand test for private `Sequences::iter()`

Expanded test for private `Sequences::iter()` to mimic old doctring

junniest committed:
Add sequence iterator test to module tests

Added sequence iterator test to module tests to make sure the public impl is tested
